File: results/0.5/Network-Time-Protocol-Version-4/network-time-protocol-version-4-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct {
    uint8_t li_vn_mode;
    uint8_t stratum;
    uint8_t poll;
    uint8_t precision;
    uint32_t root_delay;
    uint32_t root_dispersion;
    uint32_t reference_id;
    uint64_t reference_timestamp;
    uint64_t origin_timestamp;
    uint64_t receive_timestamp;
    uint64_t transmit_timestamp;
    uint8_t* extension_fields;
    size_t extension_length;
    uint8_t* mac;
    size_t mac_length;
} ntp_packet;

static HParser* init_ntp_parser(void) {
    // Basic fields
    HParser* li_vn_mode = h_bits(8, false);
    HParser* stratum = h_bits(8, false);
    HParser* poll = h_bits(8, false);
    HParser* precision = h_bits(8, false);
    HParser* root_delay = h_bits(32, false);
    HParser* root_dispersion = h_bits(32, false);
    HParser* reference_id = h_bits(32, false);
    HParser* timestamp = h_bits(64, false);

    // Extension field
    HParser* extension_field_type = h_bits(16, false);
    HParser* extension_field_length = h_bits(16, false);
    HParser* extension_field_value = h_length_value(extension_field_length, h_uint8());
    HParser* extension_field = h_sequence(extension_field_type, extension_field_length, 
                                        extension_field_value, NULL);
    HParser* extension_fields = h_many(extension_field);

    // MAC field
    HParser* key_id = h_bits(32, false);
    HParser* message_digest = h_length_value(h_int_range(h_bits(16, false), 16, 20), h_uint8());
    HParser* mac = h_sequence(key_id, message_digest, NULL);

    // Complete NTP packet
    return h_sequence(li_vn_mode,
                     stratum,
                     poll,
                     precision,
                     root_delay,
                     root_dispersion,
                     reference_id,
                     timestamp,  // reference timestamp
                     timestamp,  // origin timestamp
                     timestamp,  // receive timestamp
                     timestamp,  // transmit timestamp
                     extension_fields,
                     h_optional(mac),
                     NULL);
}

static void print_ntp_packet(const HParsedToken* token) {
    if (!token) return;

    printf("LI_VN_MODE: 0x%02x\n", (uint8_t)token->seq->elements[0]->uint);
    printf("Stratum: %u\n", (uint8_t)token->seq->elements[1]->uint);
    printf("Poll: %d\n", (int8_t)token->seq->elements[2]->uint);
    printf("Precision: %d\n", (int8_t)token->seq->elements[3]->uint);
    printf("Root Delay: 0x%08x\n", (uint32_t)token->seq->elements[4]->uint);
    printf("Root Dispersion: 0x%08x\n", (uint32_t)token->seq->elements[5]->uint);
    printf("Reference ID: 0x%08x\n", (uint32_t)token->seq->elements[6]->uint);
    printf("Reference Timestamp: 0x%016lx\n", (uint64_t)token->seq->elements[7]->uint);
    printf("Origin Timestamp: 0x%016lx\n", (uint64_t)token->seq->elements[8]->uint);
    printf("Receive Timestamp: 0x%016lx\n", (uint64_t)token->seq->elements[9]->uint);
    printf("Transmit Timestamp: 0x%016lx\n", (uint64_t)token->seq->elements[10]->uint);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <ntp_packet_file>\n", argv[0]);
        return 1;
    }

    FILE *f = fopen(argv[1], "rb");
    if (!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t *buf = malloc(size);
    if (!buf) {
        fclose(f);
        fprintf(stderr, "Memory allocation failed\n");
        return 1;
    }

    if (fread(buf, 1, size, f) != size) {
        free(buf);
        fclose(f);
        fprintf(stderr, "Failed to read file\n");
        return 1;
    }
    fclose(f);

    HParser *ntp_parser = init_ntp_parser();
    if (!ntp_parser) {
        free(buf);
        fprintf(stderr, "Failed to initialize parser\n");
        return 1;
    }

    HParseResult *result = h_parse(ntp_parser, buf, size);
    if (!result) {
        free(buf);
        fprintf(stderr, "Parse failed\n");
        return 1;
    }

    print_ntp_packet(result->ast);

    h_parse_result_free(result);
    free(buf);
    return 0;
}

File: results/0.5/Network-Time-Protocol-Version-4/network-time-protocol-version-4-claude-3-5-sonnet-20241022.rs
use nom::{
    bits::complete::take,
    combinator::map,
    error::Error,
    number::complete::{be_f32, be_i32, be_u16, be_u32, be_u8},
    sequence::tuple,
    IResult,
};
use std::{env, fs::File, io::Read};

#[derive(Debug)]
struct NTPTimestamp {
    seconds: u32,
    fraction: u32,
}

#[derive(Debug)]
struct ReferenceIdentifier {
    id: [u8; 4],
}

#[derive(Debug)]
struct NTPPacket {
    li_vn_mode: u8,
    stratum: u8,
    poll: i8,
    precision: i8,
    root_delay: f32,
    root_dispersion: f32,
    reference_id: ReferenceIdentifier,
    reference_timestamp: NTPTimestamp,
    origin_timestamp: NTPTimestamp,
    receive_timestamp: NTPTimestamp,
    transmit_timestamp: NTPTimestamp,
    key_identifier: Option<u32>,
    message_digest: Option<[u8; 16]>,
}

fn parse_timestamp(input: &[u8]) -> IResult<&[u8], NTPTimestamp> {
    let (input, (seconds, fraction)) = tuple((be_u32, be_u32))(input)?;
    Ok((
        input,
        NTPTimestamp {
            seconds,
            fraction,
        },
    ))
}

fn parse_reference_id(input: &[u8]) -> IResult<&[u8], ReferenceIdentifier> {
    let (input, id) = map(take(4usize), |bytes: &[u8]| {
        let mut arr = [0u8; 4];
        arr.copy_from_slice(bytes);
        arr
    })(input)?;
    Ok((input, ReferenceIdentifier { id }))
}

fn parse_ntp_packet(input: &[u8]) -> IResult<&[u8], NTPPacket> {
    let (input, (
        li_vn_mode,
        stratum,
        poll,
        precision,
        root_delay,
        root_dispersion,
        reference_id,
        reference_timestamp,
        origin_timestamp,
        receive_timestamp,
        transmit_timestamp,
    )) = tuple((
        be_u8,
        be_u8,
        be_i32,
        be_i32,
        be_f32,
        be_f32,
        parse_reference_id,
        parse_timestamp,
        parse_timestamp,
        parse_timestamp,
        parse_timestamp,
    ))(input)?;

    let (input, key_identifier) = if input.len() >= 4 {
        let (input, key_id) = be_u32(input)?;
        (input, Some(key_id))
    } else {
        (input, None)
    };

    let (input, message_digest) = if input.len() >= 16 {
        let (input, digest) = map(take(16usize), |bytes: &[u8]| {
            let mut arr = [0u8; 16];
            arr.copy_from_slice(bytes);
            arr
        })(input)?;
        (input, Some(digest))
    } else {
        (input, None)
    };

    Ok((
        input,
        NTPPacket {
            li_vn_mode,
            stratum,
            poll: poll as i8,
            precision: precision as i8,
            root_delay,
            root_dispersion,
            reference_id,
            reference_timestamp,
            origin_timestamp,
            receive_timestamp,
            transmit_timestamp,
            key_identifier,
            message_digest,
        },
    ))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <ntp_binary_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_ntp_packet(&buffer) {
        Ok((remaining, packet)) => {
            println!("Parsed NTP packet: {:#?}", packet);
            if !remaining.is_empty() {
                println!("Warning: {} unparsed bytes remaining", remaining.len());
            }
        }
        Err(e) => eprintln!("Failed to parse NTP packet: {:?}", e),
    }

    Ok(())
}